opt out of bottleneck for nanmean

sebasv opened the pull request:
- [x] closes #42878
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

sebasv committed:
opt out of bottleneck for nanmean

pep8speaks commented:
Hello @sebasv! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/nanops.py`](https://github.com/pandas-dev/pandas/blob/93438ec75bd943470f9ced9da834896bd87404fa/pandas/core/nanops.py):

> [Line 166:76](https://github.com/pandas-dev/pandas/blob/93438ec75bd943470f9ced9da834896bd87404fa/pandas/core/nanops.py#L166): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

sebasv committed:
remove trailing whitespace

sebasv committed:
make error bound explicit

sebasv commented:
I added 2 unit tests to validate that numerical precision is bounded by log(n). If anyone has suggestions how to improve these tests let me know

mroeschke commented on pandas/tests/reductions/test_reductions.py:
```diff
@@ -1534,3 +1534,30 @@ def test_multimode_complex(self, array, expected, dtype):
         tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize("dtype", ["float32", "float64"])
```
I think instead of these tests, might be better to test in `pandas/test/test_nanops` that bottleneck is not used i.e. test `_bn_ok_dtype` doesn't allow mean.

IMO these tests are essentially testing numpy behavior which ideally numpy should testing the precision instead.

sebasv replied:
Fair point, I guess it's quite safe to trust numpy have a reasonably precise implementation of `sum`. I'll update

sebasv committed:
unittest only _bn_ok_dtype

mroeschke commented on pandas/tests/test_nanops.py:
```diff
@@ -1120,3 +1120,9 @@ def test_check_below_min_count__large_shape(min_count, expected_result):
+
+@pytest.mark.parametrize("func", ["nanmean", "nansum"])
+@pytest.mark.parametrize("dtype", [np.float16, np.float32, np.float64])
+def test_check_bottleneck_disallow(dtype, func):
```
Could you add `# GH 42878` as a comment here?

sebasv committed:
link issue to test function

mroeschke commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -844,7 +844,7 @@ Numeric
 - Bug in division, ``pow`` and ``mod`` operations on array-likes with ``dtype="boolean"`` not being like their ``np.bool_`` counterparts (:issue:`46063`)
 - Bug in multiplying a :class:`Series` with ``IntegerDtype`` or ``FloatingDtype`` by an array-like with ``timedelta64[ns]`` dtype incorrectly raising (:issue:`45622`)
--
+- Bug in :meth:`mean` where the optional dependency ``bottleneck`` causes precision loss linear in the length of the array. By falling back to numpy the loss is now log-linear (:issue:`42878`)
```
```suggestion
- Bug in :meth:`mean` where the optional dependency ``bottleneck`` causes precision loss linear in the length of the array. ``bottleneck`` has been disabled for :meth:`mean` improving the loss to log-linear but may result in a performance decrease. (:issue:`42878`)
```

mroeschke reviewed (commented):
Confirming that this should be disabled for integer types too? i.e. the precision loss is similar for integers and therefore should also be disabled?

sebasv committed:
Update doc/source/whatsnew/v1.5.0.rst

clarify that there might be a performance decrease experienced from disabling `mean` for bottleneck

Co-authored-by: Matthew Roeschke <emailformattr@gmail.com>

sebasv commented:
@mroeschke I will verify this. Addition is pure for integers so depending on the bottleneck implementation it should be almost exact

JMBurley commented:
Glad to see this update.  Should it be targeting pandas 1.5.0 or 1.4.4 release?

mroeschke commented:
> Glad to see this update. Should it be targeting pandas 1.5.0 or 1.4.4 release?

1.5. Point releases are reserved for regression in behavior from recent releases.

sebasv commented:
Integer mean should also be opted out of. In the case of integer arrays it seems that bottleneck uses a float64 to hold the aggregating sum, wich means that the problem persists there as well. I'm adding int types to the unit test as well

sebasv committed:
extend unit tests with (u)int dtypes

mroeschke commented on pandas/core/nanops.py:
```diff
@@ -162,6 +162,9 @@ def f(
     if not is_object_dtype(dtype) and not needs_i8_conversion(dtype):
+        # GH 42878
+        # Bottleneck uses naive summation leading to O(n) loss of precision
+        # unlike numpy which implements pairwise summation, which has O(log(n)) loss
```
One more item: Is there a issue on the bottleneck Github about the precision issue that you can comment here?

JMBurley replied:
I believe pydata/bottleneck/issues/379 is the pertinent issue

JMBurley commented on pandas/core/nanops.py:
```diff
@@ -162,6 +162,9 @@ def f(
     if not is_object_dtype(dtype) and not needs_i8_conversion(dtype):
+        # GH 42878
+        # Bottleneck uses naive summation leading to O(n) loss of precision
+        # unlike numpy which implements pairwise summation, which has O(log(n)) loss
```
Would this be the preferred way to link? Or is there another  standard syntax for how pandas writes such comments?
```suggestion
        # unlike numpy which implements pairwise summation, which has O(log(n)) loss
        # crossref: https://github.com/pydata/bottleneck/issues/379
```

sebasv committed:
Update pandas/core/nanops.py

Co-authored-by: JMBurley <JMBurley@users.noreply.github.com>

mroeschke commented:
Thanks @sebasv and @JMBurley for review
